File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/objects/TransactionObjectTest.java
package com.github.dedis.popstellar.model.objects;

import com.github.dedis.popstellar.model.network.JsonTestUtils;
import com.github.dedis.popstellar.model.network.method.message.data.digitalcash.*;
import com.github.dedis.popstellar.model.objects.digitalcash.*;
import com.github.dedis.popstellar.model.objects.security.*;
import com.github.dedis.popstellar.utility.security.Hash;

import org.junit.Test;

import java.security.GeneralSecurityException;
import java.util.*;

import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateKeyPair;
import static org.junit.Assert.*;

public class TransactionObjectTest {
  KeyPair senderKey = generateKeyPair();
  PublicKey sender = senderKey.getPublicKey();

  @Test
  public void getChannelTest() throws GeneralSecurityException {
    Channel channel = Channel.fromString("/root/laoId/coin/myChannel");
    TransactionObjectBuilder builder = getValidTransactionBuilder();
    builder.setChannel(channel);
    assertEquals(channel, builder.build().getChannel());
  }

  // test get Inputs
  @Test
  public void getInputsTest() throws GeneralSecurityException {
    TransactionObjectBuilder builder = getValidTransactionBuilder();
    int txOutIndex = 0;
    String txOutHash = "47DEQpj8HBSa--TImW-5JCeuQeRkm5NMpJWZG3hSuFU=";
    String type = "P2PKH";

    ScriptInputObject scriptTxIn;
    String sig = senderKey.sign(sender).getEncoded();
    scriptTxIn = new ScriptInputObject(type, sender, new Signature(sig));
    InputObject input = new InputObject(txOutHash, txOutIndex, scriptTxIn);
    List<InputObject> listInput = Collections.singletonList(input);
    builder.setInputs(listInput);

    assertEquals(listInput, builder.build().getInputs());
  }

  // test get Outputs
  @Test
  public void getOutputsTest() throws GeneralSecurityException {
    TransactionObjectBuilder builder = getValidTransactionBuilder();
    String type = "P2PKH";
    String pubKeyHash = sender.computeHash();
    ScriptOutputObject scriptTxOut = new ScriptOutputObject(type, pubKeyHash);
    int value = 32;
    OutputObject output = new OutputObject(value, scriptTxOut);
    List<OutputObject> listOutput = Collections.singletonList(output);
    builder.setOutputs(listOutput);
    assertEquals(listOutput, builder.build().getOutputs());
  }

  @Test
  public void getLockTimeTest() throws GeneralSecurityException {
    TransactionObjectBuilder builder = getValidTransactionBuilder();
    long locktime = 0;
    builder.setLockTime(locktime);
    assertEquals(locktime, builder.build().getLockTime());
  }

  @Test
  public void getVersionTest() throws GeneralSecurityException {
    TransactionObjectBuilder builder = getValidTransactionBuilder();
    int version = 0;
    builder.setVersion(version);
    assertEquals(version, builder.build().getVersion());
  }

  @Test
  public void getSendersTransactionTest() throws GeneralSecurityException {
    TransactionObjectBuilder builder = getValidTransactionBuilder();
    int txOutIndex = 0;
    String txOutHash = "47DEQpj8HBSa--TImW-5JCeuQeRkm5NMpJWZG3hSuFU=";

    String type = "P2PKH";
    String pubKey = sender.getEncoded();
    ScriptInputObject scriptTxIn;
    String sig = "dhfqkdfhqu";
    scriptTxIn = new ScriptInputObject(type, new PublicKey(pubKey), new Signature(sig));
    InputObject input = new InputObject(txOutHash, txOutIndex, scriptTxIn);
    List<InputObject> listInput = Collections.singletonList(input);
    builder.setInputs(listInput);
    assertEquals(Collections.singletonList(sender), builder.build().getSendersTransaction());
  }

  @Test
  public void getReceiversHashTransactionTest() throws GeneralSecurityException {
    TransactionObjectBuilder builder = getValidTransactionBuilder();

    String type = "P2PKH";
    String pubKeyHash = sender.computeHash();
    ScriptOutputObject scriptTxOut = new ScriptOutputObject(type, pubKeyHash);
    int value = 32;
    OutputObject output = new OutputObject(value, scriptTxOut);
    List<OutputObject> listOutput = Collections.singletonList(output);
    builder.setOutputs(listOutput);
    assertEquals(
        Collections.singletonList(pubKeyHash), builder.build().getReceiversHashTransaction());
  }

  // test thrown null List<PublicKey> getReceiversTransaction(Map<String, PublicKey> mapHashKey)
  // test thrown null Map<PublicKey, Long> getReceiversTransactionMap(Map<String, PublicKey>
  // mapHashKey)
  @Test
  public void getReceiversTransactionTestNull() throws GeneralSecurityException {
    TransactionObjectBuilder builder = getValidTransactionBuilder();
    KeyPair senderKey1 = generateKeyPair();
    PublicKey sender1 = senderKey1.getPublicKey();
    PublicKey sender2 = null;
    String type = "P2PKH";
    String pubkeyhash1 = sender1.computeHash();
    String pubkeyhash2 = "none";
    ScriptOutputObject scriptTxOut = new ScriptOutputObject(type, pubkeyhash1);
    int value = 32;
    OutputObject output = new OutputObject(value, scriptTxOut);
    List<OutputObject> listOutput = Collections.singletonList(output);
    Map<String, PublicKey> mapHash = Collections.singletonMap(pubkeyhash2, sender2);
    builder.setOutputs(listOutput);
    TransactionObject transactionObject = builder.build();
    assertThrows(
        IllegalArgumentException.class, () -> transactionObject.getReceiversTransaction(mapHash));
  }

  @Test
  public void getReceiversTransactionTest() throws GeneralSecurityException {
    TransactionObjectBuilder builder = getValidTransactionBuilder();
    String type = "P2PKH";
    String pubkeyhash = sender.computeHash();
    ScriptOutputObject scriptTxOut = new ScriptOutputObject(type, pubkeyhash);
    int value = 32;
    OutputObject output = new OutputObject(value, scriptTxOut);
    List<OutputObject> listOutput = Collections.singletonList(output);
    Map<String, PublicKey> mapHash = Collections.singletonMap(pubkeyhash, sender);
    builder.setOutputs(listOutput);
    assertEquals(
        Collections.singletonList(sender), builder.build().getReceiversTransaction(mapHash));
  }

  @Test
  public void getIsReceiverTest() throws GeneralSecurityException {
    TransactionObjectBuilder builder = getValidTransactionBuilder();

    String type = "P2PKH";
    String pubKeyHash = sender.computeHash();
    ScriptOutputObject scriptTxOut = new ScriptOutputObject(type, pubKeyHash);
    int value = 32;
    OutputObject output = new OutputObject(value, scriptTxOut);
    List<OutputObject> listOutput = Collections.singletonList(output);
    builder.setOutputs(listOutput);
    TransactionObject transactionObject = builder.build();
    assertEquals(
        Collections.singletonList(pubKeyHash), transactionObject.getReceiversHashTransaction());
    // DUMMY SENDER
    KeyPair senderDummyKey = generateKeyPair();
    PublicKey senderDummy = senderDummyKey.getPublicKey();
    assertNotEquals(
        Collections.singletonList(senderDummy.computeHash()),
        transactionObject.getReceiversHashTransaction());
  }

  // test boolean is Sender(PublicKey publicKey)
  @Test
  public void getIsSenderTest() throws GeneralSecurityException {
    TransactionObjectBuilder builder = getValidTransactionBuilder();
    String type = "P2PKH";
    ScriptInputObject scriptTxInput = new ScriptInputObject(type, sender, new Signature("qqchose"));
    InputObject input = new InputObject(Hash.hash("none"), 0, scriptTxInput);
    List<InputObject> listInput = Collections.singletonList(input);
    builder.setInputs(listInput);
    assertTrue(builder.build().isSender(sender));
  }

  // test int get_index_transaction(PublicKey publicKey)
  @Test
  public void getIndexTransactionTest() throws GeneralSecurityException {
    TransactionObjectBuilder builder = getValidTransactionBuilder();

    String type = "P2PKH";
    // RECEIVER
    String pubKeyHash = sender.computeHash();
    ScriptOutputObject scriptTxOut = new ScriptOutputObject(type, pubKeyHash);
    int value = 32;
    OutputObject output = new OutputObject(value, scriptTxOut);
    List<OutputObject> listOutput = Collections.singletonList(output);
    builder.setOutputs(listOutput);
    assertEquals(0, builder.build().getIndexTransaction(sender));
  }

  @Test
  public void computeIdTest() throws GeneralSecurityException {
    String path = "protocol/examples/messageData/coin/post_transaction_coinbase.json";
    String validJson = JsonTestUtils.loadFile(path);
    PostTransactionCoin postTransactionModel = (PostTransactionCoin) JsonTestUtils.parse(validJson);
    Transaction transactionModel = postTransactionModel.getTransaction();

    TransactionObjectBuilder builder = getValidTransactionBuilder();
    builder.setLockTime(transactionModel.getLockTime());
    builder.setVersion(transactionModel.getVersion());

    List<InputObject> inpObj = new ArrayList<>();
    List<OutputObject> outObj = new ArrayList<>();
    for (Input i : transactionModel.getInputs()) {
      ScriptInput scriptInput = i.getScript();
      inpObj.add(
          new InputObject(
              i.getTxOutHash(),
              i.getTxOutIndex(),
              new ScriptInputObject(
                  scriptInput.getType(), scriptInput.getPubkey(), scriptInput.getSig())));
    }

    for (Output o : transactionModel.getOutputs()) {
      ScriptOutput scriptOutput = o.getScript();
      outObj.add(
          new OutputObject(
              o.getValue(),
              new ScriptOutputObject(scriptOutput.getType(), scriptOutput.getPubKeyHash())));
    }

    builder.setInputs(inpObj);
    builder.setOutputs(outObj);

    TransactionObject transactionObject = builder.build();
    assertTrue(transactionObject.isCoinBaseTransaction());
  }

  private TransactionObjectBuilder getValidTransactionBuilder() throws GeneralSecurityException {
    return getValidTransactionBuilder("a");
  }

  private TransactionObjectBuilder getValidTransactionBuilder(String transactionId)
      throws GeneralSecurityException {
    TransactionObjectBuilder builder = new TransactionObjectBuilder();

    String type = "P2PKH";
    int txOutIndex = 0;
    String txOutHash = "47DEQpj8HBSa--TImW-5JCeuQeRkm5NMpJWZG3hSuFU=";

    ScriptInputObject scriptTxIn;
    String sig = senderKey.sign(sender).getEncoded();
    scriptTxIn = new ScriptInputObject(type, sender, new Signature(sig));
    InputObject input = new InputObject(txOutHash, txOutIndex, scriptTxIn);
    List<InputObject> listInput = Collections.singletonList(input);
    builder.setInputs(listInput);

    Channel channel = Channel.fromString("/root/laoId/coin/myChannel");
    builder.setChannel(channel);

    String pubKeyHash = sender.computeHash();
    ScriptOutputObject scriptTxOut = new ScriptOutputObject(type, pubKeyHash);
    int value = 32;
    OutputObject output = new OutputObject(value, scriptTxOut);
    List<OutputObject> listOutput = Collections.singletonList(output);
    builder.setOutputs(listOutput);

    builder.setLockTime(0);
    builder.setVersion(0);
    builder.setTransactionId(transactionId);
    return builder;
  }
}

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/objects/TransactionObjectTest.kt
package com.github.dedis.popstellar.model.objects

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.github.dedis.popstellar.model.network.JsonTestUtils.loadFile
import com.github.dedis.popstellar.model.network.JsonTestUtils.parse
import com.github.dedis.popstellar.model.network.method.message.data.digitalcash.PostTransactionCoin
import com.github.dedis.popstellar.model.objects.Channel.Companion.fromString
import com.github.dedis.popstellar.model.objects.digitalcash.ScriptInputObject
import com.github.dedis.popstellar.model.objects.digitalcash.ScriptOutputObject
import com.github.dedis.popstellar.model.objects.digitalcash.TransactionObjectBuilder
import com.github.dedis.popstellar.model.objects.security.KeyPair
import com.github.dedis.popstellar.model.objects.security.PublicKey
import com.github.dedis.popstellar.model.objects.security.Signature
import com.github.dedis.popstellar.testutils.Base64DataUtils
import com.github.dedis.popstellar.utility.security.HashSHA256.hash
import java.security.GeneralSecurityException
import java.util.Collections
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class TransactionObjectTest {
  private var senderKey: KeyPair = Base64DataUtils.generateKeyPair()
  var sender = senderKey.publicKey

  @Throws(GeneralSecurityException::class)
  @Test
  fun channelTest() {
    val channel = fromString("/root/laoId/coin/myChannel")
    val builder = validTransactionBuilder
    builder.setChannel(channel)
    Assert.assertEquals(channel, builder.build().channel)
  }

  @Throws(GeneralSecurityException::class)
  @Test
  fun inputsTest() {
    // test get Inputs
    val builder = validTransactionBuilder
    val txOutIndex = 0
    val txOutHash = "47DEQpj8HBSa--TImW-5JCeuQeRkm5NMpJWZG3hSuFU="
    val type = "P2PKH"
    val scriptTxIn: ScriptInputObject
    val sig = senderKey.sign(sender).encoded
    scriptTxIn = ScriptInputObject(type, sender, Signature(sig))
    val input = InputObject(txOutHash, txOutIndex, scriptTxIn)
    val listInput = listOf(input)
    builder.setInputs(listInput)

    Assert.assertEquals(listInput, builder.build().inputs)
  }

  @Throws(GeneralSecurityException::class)
  @Test
  fun outputsTest() {
    // test get Outputs
    val builder = validTransactionBuilder
    val type = "P2PKH"
    val pubKeyHash = sender.computeHash()
    val scriptTxOut = ScriptOutputObject(type, pubKeyHash)
    val value = 32
    val output = OutputObject(value.toLong(), scriptTxOut)
    val listOutput = listOf(output)
    builder.setOutputs(listOutput)

    Assert.assertEquals(listOutput, builder.build().outputs)
  }

  @Throws(GeneralSecurityException::class)
  @Test
  fun lockTimeTest() {
    val builder = validTransactionBuilder
    val locktime: Long = 0
    builder.setLockTime(locktime)
    Assert.assertEquals(locktime, builder.build().lockTime)
  }

  @Throws(GeneralSecurityException::class)
  @Test
  fun versionTest() {
    val builder = validTransactionBuilder
    val version = 0
    builder.setVersion(version)
    Assert.assertEquals(version.toLong(), builder.build().version.toLong())
  }

  @Throws(GeneralSecurityException::class)
  @Test
  fun sendersTransactionTest() {
    val builder = validTransactionBuilder
    val txOutIndex = 0
    val txOutHash = "47DEQpj8HBSa--TImW-5JCeuQeRkm5NMpJWZG3hSuFU="
    val type = "P2PKH"
    val pubKey = sender.encoded
    val scriptTxIn: ScriptInputObject
    val sig = "dhfqkdfhqu"
    scriptTxIn = ScriptInputObject(type, PublicKey(pubKey), Signature(sig))
    val input = InputObject(txOutHash, txOutIndex, scriptTxIn)
    val listInput = listOf(input)
    builder.setInputs(listInput)

    Assert.assertEquals(listOf(sender), builder.build().sendersTransaction)
  }

  @Throws(GeneralSecurityException::class)
  @Test
  fun receiversHashTransactionTest() {
    val builder = validTransactionBuilder
    val type = "P2PKH"
    val pubKeyHash = sender.computeHash()
    val scriptTxOut = ScriptOutputObject(type, pubKeyHash)
    val value = 32
    val output = OutputObject(value.toLong(), scriptTxOut)
    val listOutput = listOf(output)
    builder.setOutputs(listOutput)

    Assert.assertEquals(listOf(pubKeyHash), builder.build().receiversHashTransaction)
  }

  @Throws(GeneralSecurityException::class)
  @Test
  fun receiversTransactionTestNull() {
    // test thrown null List<PublicKey> getReceiversTransaction(Map<String, PublicKey> mapHashKey)
    val builder = validTransactionBuilder
    val senderKey1 = Base64DataUtils.generateKeyPair()
    val sender1 = senderKey1.publicKey
    val sender2: PublicKey? = null
    val type = "P2PKH"
    val pubkeyhash1 = sender1.computeHash()
    val pubkeyhash2 = "none"
    val scriptTxOut = ScriptOutputObject(type, pubkeyhash1)
    val value = 32
    val output = OutputObject(value.toLong(), scriptTxOut)
    val listOutput = listOf(output)
    val mapHash = Collections.singletonMap(pubkeyhash2, sender2)
    builder.setOutputs(listOutput)
    val transactionObject = builder.build()
    Assert.assertThrows(IllegalArgumentException::class.java) {
      transactionObject.getReceiversTransaction(mapHash.toMap())
    }
  }

  @Throws(GeneralSecurityException::class)
  @Test
  fun receiversTransactionTest() {
    val builder = validTransactionBuilder
    val type = "P2PKH"
    val pubkeyhash = sender.computeHash()
    val scriptTxOut = ScriptOutputObject(type, pubkeyhash)
    val value = 32
    val output = OutputObject(value.toLong(), scriptTxOut)
    val listOutput = listOf(output)
    val mapHash = Collections.singletonMap(pubkeyhash, sender)
    builder.setOutputs(listOutput)
    Assert.assertEquals(listOf(sender), builder.build().getReceiversTransaction(mapHash))
  }

  @Throws(GeneralSecurityException::class)
  @Test
  fun isReceiverTest() {
    val builder = validTransactionBuilder
    val type = "P2PKH"
    val pubKeyHash = sender.computeHash()
    val scriptTxOut = ScriptOutputObject(type, pubKeyHash)
    val value = 32
    val output = OutputObject(value.toLong(), scriptTxOut)
    val listOutput = listOf(output)
    builder.setOutputs(listOutput)
    val transactionObject = builder.build()
    Assert.assertEquals(listOf(pubKeyHash), transactionObject.receiversHashTransaction)
    // DUMMY SENDER
    val senderDummyKey = Base64DataUtils.generateKeyPair()
    val senderDummy = senderDummyKey.publicKey
    Assert.assertNotEquals(
      listOf(senderDummy.computeHash()),
      transactionObject.receiversHashTransaction
    )
  }

  @Throws(GeneralSecurityException::class)
  @Test
  fun isSenderTest() {
    // test boolean is Sender(PublicKey publicKey)
    val builder = validTransactionBuilder
    val type = "P2PKH"
    val scriptTxInput = ScriptInputObject(type, sender, Signature("qqchose"))
    val input = InputObject(hash("none"), 0, scriptTxInput)
    val listInput = listOf(input)
    builder.setInputs(listInput)
    Assert.assertTrue(builder.build().isSender(sender))
  }

  @Throws(GeneralSecurityException::class)
  @Test
  fun indexTransactionTest() {
    // test int get_index_transaction(PublicKey publicKey)
    val builder = validTransactionBuilder
    val type = "P2PKH"
    // RECEIVER
    val pubKeyHash = sender.computeHash()
    val scriptTxOut = ScriptOutputObject(type, pubKeyHash)
    val value = 32
    val output = OutputObject(value.toLong(), scriptTxOut)
    val listOutput = listOf(output)
    builder.setOutputs(listOutput)
    Assert.assertEquals(0, builder.build().getIndexTransaction(sender).toLong())
  }

  @Test
  @Throws(GeneralSecurityException::class)
  fun computeIdTest() {
    val path = "protocol/examples/messageData/coin/post_transaction_coinbase.json"
    val validJson = loadFile(path)
    val postTransactionModel = parse(validJson) as PostTransactionCoin
    val transactionModel = postTransactionModel.transaction
    val builder = validTransactionBuilder
    builder.setLockTime(transactionModel.lockTime)
    builder.setVersion(transactionModel.version)
    val inpObj: MutableList<InputObject> = ArrayList()
    val outObj: MutableList<OutputObject> = ArrayList()
    for (i in transactionModel.inputs) {
      val scriptInput = i.script
      inpObj.add(
        InputObject(
          i.txOutHash,
          i.txOutIndex,
          ScriptInputObject(scriptInput.type, scriptInput.pubkey, scriptInput.sig)
        )
      )
    }
    for (o in transactionModel.outputs) {
      val scriptOutput = o.script
      outObj.add(
        OutputObject(o.value, ScriptOutputObject(scriptOutput.type, scriptOutput.pubKeyHash))
      )
    }
    builder.setInputs(inpObj)
    builder.setOutputs(outObj)
    val transactionObject = builder.build()
    Assert.assertTrue(transactionObject.isCoinBaseTransaction)
  }

  @get:Throws(GeneralSecurityException::class)
  private val validTransactionBuilder: TransactionObjectBuilder
    get() = getValidTransactionBuilder("a")

  @Throws(GeneralSecurityException::class)
  private fun getValidTransactionBuilder(transactionId: String): TransactionObjectBuilder {
    val builder = TransactionObjectBuilder()

    val type = "P2PKH"
    val txOutIndex = 0
    val txOutHash = "47DEQpj8HBSa--TImW-5JCeuQeRkm5NMpJWZG3hSuFU="
    val scriptTxIn: ScriptInputObject
    val sig = senderKey.sign(sender).encoded
    scriptTxIn = ScriptInputObject(type, sender, Signature(sig))
    val input = InputObject(txOutHash, txOutIndex, scriptTxIn)
    val listInput = listOf(input)
    builder.setInputs(listInput)

    val channel = fromString("/root/laoId/coin/myChannel")
    builder.setChannel(channel)

    val pubKeyHash = sender.computeHash()
    val scriptTxOut = ScriptOutputObject(type, pubKeyHash)
    val value = 32
    val output = OutputObject(value.toLong(), scriptTxOut)
    val listOutput = listOf(output)

    builder.setOutputs(listOutput)
    builder.setLockTime(0)
    builder.setVersion(0)
    builder.setTransactionId(transactionId)

    return builder
  }
}
